File: server/src/ws-pty.ts
import * as http from 'http';
import { WebSocketServer, WebSocket } from 'ws';
import { URL } from 'url';
import { ptyManager } from './pty-manager';

export function attachPtyWebSockets(server: http.Server): void {
    const wss = new WebSocketServer({ noServer: true });

    server.on('upgrade', (req, socket, head) => {
        const pathname = new URL(req.url ?? '', `http://${req.headers.host}`).pathname;

        if (pathname === '/services/pty/control') {
            wss.handleUpgrade(req, socket, head, (ws) => handleControl(ws));
            return;
        }

        const m = pathname.match(/^\/pty\/([^/]+)\/(data|resize|buffer-watch)$/);
        if (m) {
            const [, instanceId, channel] = m;
            wss.handleUpgrade(req, socket, head, (ws) => {
                if (channel === 'data') handleData(ws, instanceId);
                else if (channel === 'resize') handleResize(ws, instanceId);
                else handleBufferWatch(ws, instanceId);
            });
        }
    });
}

function handleControl(ws: WebSocket): void {
    ws.on('message', (raw) => {
        try {
            const msg = JSON.parse(raw.toString());
            if (msg.type === 'spawn') {
                ptyManager.spawn(msg.instanceId, msg.role, msg.cols ?? 120, msg.rows ?? 30, msg.cwd);
                ws.send(JSON.stringify({ instanceId: msg.instanceId, status: 'ok' }));
            } else if (msg.type === 'kill') {
                ptyManager.kill(msg.instanceId);
                ws.send(JSON.stringify({ instanceId: msg.instanceId, status: 'ok' }));
            }
        } catch { /* ignore */ }
    });
}

function handleData(ws: WebSocket, instanceId: string): void {
    const unsub = ptyManager.onData(instanceId, (data) => {
        if (ws.readyState === WebSocket.OPEN) ws.send(data);
    });
    ws.on('message', (raw) => {
        const data = raw instanceof Buffer ? raw : Buffer.from(raw as ArrayBuffer);
        ptyManager.write(instanceId, data);
    });
    ws.on('close', () => unsub());
}

function handleResize(ws: WebSocket, instanceId: string): void {
    ws.on('message', (raw) => {
        const [cols, rows] = raw.toString().split(',').map(Number);
        if (cols > 0 && rows > 0) ptyManager.resize(instanceId, cols, rows);
    });
}

function handleBufferWatch(ws: WebSocket, instanceId: string): void {
    let last = '';
    let inFlight = false;
    const interval = setInterval(async () => {
        if (inFlight || ws.readyState !== WebSocket.OPEN) return;
        inFlight = true;
        try {
            const sockPath = ptyManager.getSocketPath(instanceId);
            if (!sockPath) return;
            const { execFile } = await import('child_process');
            const file = await new Promise<string>((res, rej) =>
                execFile(ptyManager.getNvimBin(), ['--server', sockPath, '--remote-expr', 'expand("%:p")'],
                    { timeout: 300 }, (err, out) => err ? rej(err) : res(out.trim()))
            );
            if (file && file !== last) { last = file; ws.send(file); }
        } catch { /* nvim not ready */ }
        finally { inFlight = false; }
    }, 500);
    ws.on('close', () => clearInterval(interval));
}

File: server/ws_pty.go
package main

import (
	"context"
	"encoding/json"
	"log"
	"net/http"
	"os/exec"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/gorilla/websocket"
)

// ptyPathRe matches /pty/<instanceId>/<channel>
var ptyPathRe = regexp.MustCompile(`^/pty/([^/]+)/(data|resize|buffer-watch)$`)

// RegisterPtyWebSockets registers all PTY-related WebSocket upgrade handlers.
func RegisterPtyWebSockets(mux *http.ServeMux, upgrader *websocket.Upgrader, ptyMgr *PtyManager) {
	mux.HandleFunc("/services/pty/control", func(w http.ResponseWriter, r *http.Request) {
		conn, err := upgrader.Upgrade(w, r, nil)
		if err != nil {
			log.Printf("[ws-pty] control upgrade failed: %v", err)
			return
		}
		handleControl(conn, ptyMgr)
	})

	// Dynamic PTY paths: /pty/{id}/{channel}
	// We register a broad prefix and parse the path ourselves.
	mux.HandleFunc("/pty/", func(w http.ResponseWriter, r *http.Request) {
		m := ptyPathRe.FindStringSubmatch(r.URL.Path)
		if m == nil {
			http.NotFound(w, r)
			return
		}
		instanceID := m[1]
		channel := m[2]

		conn, err := upgrader.Upgrade(w, r, nil)
		if err != nil {
			log.Printf("[ws-pty] %s upgrade failed: %v", channel, err)
			return
		}

		switch channel {
		case "data":
			handleData(conn, instanceID, ptyMgr)
		case "resize":
			handleResize(conn, instanceID, ptyMgr)
		case "buffer-watch":
			handleBufferWatch(conn, instanceID, ptyMgr)
		}
	})
}

// controlMsg is the JSON structure for control channel messages.
type controlMsg struct {
	Type       string `json:"type"`
	InstanceID string `json:"instanceId"`
	Role       string `json:"role"`
	Cols       int    `json:"cols"`
	Rows       int    `json:"rows"`
	Cwd        string `json:"cwd"`
}

// controlReply is sent back on the control channel.
type controlReply struct {
	InstanceID string `json:"instanceId"`
	Status     string `json:"status"`
	Error      string `json:"error,omitempty"`
}

func handleControl(conn *websocket.Conn, ptyMgr *PtyManager) {
	defer conn.Close()

	for {
		_, raw, err := conn.ReadMessage()
		if err != nil {
			return
		}

		var msg controlMsg
		if err := json.Unmarshal(raw, &msg); err != nil {
			continue
		}

		cols := msg.Cols
		if cols <= 0 {
			cols = 120
		}
		rows := msg.Rows
		if rows <= 0 {
			rows = 30
		}

		switch msg.Type {
		case "spawn":
			role := PaneRole(msg.Role)
			if role != RoleNeovim && role != RoleTerminal {
				role = RoleTerminal
			}
			spawnErr := ptyMgr.Spawn(msg.InstanceID, role, uint16(cols), uint16(rows), msg.Cwd)
			reply := controlReply{InstanceID: msg.InstanceID, Status: "ok"}
			if spawnErr != nil {
				reply.Status = "error"
				reply.Error = spawnErr.Error()
			}
			data, _ := json.Marshal(reply)
			conn.WriteMessage(websocket.TextMessage, data)

		case "kill":
			ptyMgr.Kill(msg.InstanceID)
			reply := controlReply{InstanceID: msg.InstanceID, Status: "ok"}
			data, _ := json.Marshal(reply)
			conn.WriteMessage(websocket.TextMessage, data)
		}
	}
}

func handleData(conn *websocket.Conn, instanceID string, ptyMgr *PtyManager) {
	// Subscribe to PTY output and forward to WebSocket
	unsub := ptyMgr.OnData(instanceID, func(data []byte) {
		conn.WriteMessage(websocket.BinaryMessage, data)
	})
	defer unsub()
	defer conn.Close()

	// Read from WebSocket and write to PTY stdin
	for {
		_, raw, err := conn.ReadMessage()
		if err != nil {
			return
		}
		ptyMgr.Write(instanceID, raw)
	}
}

func handleResize(conn *websocket.Conn, instanceID string, ptyMgr *PtyManager) {
	defer conn.Close()

	for {
		_, raw, err := conn.ReadMessage()
		if err != nil {
			return
		}
		parts := strings.SplitN(string(raw), ",", 2)
		if len(parts) != 2 {
			continue
		}
		cols, err1 := strconv.Atoi(strings.TrimSpace(parts[0]))
		rows, err2 := strconv.Atoi(strings.TrimSpace(parts[1]))
		if err1 != nil || err2 != nil || cols <= 0 || rows <= 0 {
			continue
		}
		ptyMgr.Resize(instanceID, uint16(cols), uint16(rows))
	}
}

func handleBufferWatch(conn *websocket.Conn, instanceID string, ptyMgr *PtyManager) {
	defer conn.Close()

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	// Cancel context when connection closes
	go func() {
		for {
			_, _, err := conn.ReadMessage()
			if err != nil {
				cancel()
				return
			}
		}
	}()

	last := ""
	ticker := time.NewTicker(500 * time.Millisecond)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			sockPath := ptyMgr.GetSocketPath(instanceID)
			if sockPath == "" {
				continue
			}

			nvimBin := ptyMgr.GetNvimBin()
			cmdCtx, cmdCancel := context.WithTimeout(ctx, 300*time.Millisecond)
			out, err := exec.CommandContext(cmdCtx, nvimBin,
				"--server", sockPath,
				"--remote-expr", `expand("%:p")`).Output()
			cmdCancel()

			if err != nil {
				continue
			}

			file := strings.TrimSpace(string(out))
			if file != "" && file != last {
				last = file
				if err := conn.WriteMessage(websocket.TextMessage, []byte(file)); err != nil {
					return
				}
			}
		}
	}
}
